<a href="https://colab.research.google.com/github/aryasovaalexandra/praktika/blob/main/%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5!1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/content/crimes_against_women_2001-2014.csv')
df

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
0               0  ANDHRA PRADESH           ADILABAD  2001    50   
1               1  ANDHRA PRADESH          ANANTAPUR  2001    23   
2               2  ANDHRA PRADESH           CHITTOOR  2001    27   
3               3  ANDHRA PRADESH           CUDDAPAH  2001    20   
4               4  ANDHRA PRADESH      EAST GODAVARI  2001    23   
...           ...             ...                ...   ...   ...   
10672         832     Lakshadweep        Lakshadweep  2014     1   
10673         833     Lakshadweep  Total District(s)  2014     1   
10674         834      Puducherry           Karaikal  2014     3   
10675         835      Puducherry         Puducherry  2014     7   
10676         836      Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
0                            30            16   
1                            30             7   
2                            34            14   
3                            20            17   
4                            26            12   
...                         ...           ...   
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
0                                                    149     
1                                                    118     
2                                                    112     
3                                                    126     
4                                                    109     
...                                                  ...     
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                              34                                  175   
1                              24                                  154   
2                              83                                  186   
3                              38                                   57   
4                              58                                  247   
...                           ...                                  ...   
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  
...                     ...  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0  

[10677 rows x 11 columns]

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [ ]:
df.columns


Index(['Unnamed: 0', 'STATE/UT', 'DISTRICT', 'Year', 'Rape',
       'Kidnapping and Abduction', 'Dowry Deaths',
       'Assault on women with intent to outrage her modesty',
       'Insult to modesty of Women', 'Cruelty by Husband or his Relatives',
       'Importation of Girls'],
      dtype='object')

In [ ]:
df.head(5)

Unnamed: 0        STATE/UT       DISTRICT  Year  Rape  \
0           0  ANDHRA PRADESH       ADILABAD  2001    50   
1           1  ANDHRA PRADESH      ANANTAPUR  2001    23   
2           2  ANDHRA PRADESH       CHITTOOR  2001    27   
3           3  ANDHRA PRADESH       CUDDAPAH  2001    20   
4           4  ANDHRA PRADESH  EAST GODAVARI  2001    23   

   Kidnapping and Abduction  Dowry Deaths  \
0                        30            16   
1                        30             7   
2                        34            14   
3                        20            17   
4                        26            12   

   Assault on women with intent to outrage her modesty  \
0                                                149     
1                                                118     
2                                                112     
3                                                126     
4                                                109     

   Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                          34                                  175   
1                          24                                  154   
2                          83                                  186   
3                          38                                   57   
4                          58                                  247   

   Importation of Girls  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0

In [ ]:
df.tail(5)

Unnamed: 0     STATE/UT           DISTRICT  Year  Rape  \
10672         832  Lakshadweep        Lakshadweep  2014     1   
10673         833  Lakshadweep  Total District(s)  2014     1   
10674         834   Puducherry           Karaikal  2014     3   
10675         835   Puducherry         Puducherry  2014     7   
10676         836   Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0

In [ ]:
df.describe()

Unnamed: 0          Year          Rape  Kidnapping and Abduction  \
count  10677.000000  10677.000000  10677.000000              10677.000000   
mean    3871.570198   2007.697949     57.989885                 69.888358   
std     2816.369918      4.046874    214.230398                311.623450   
min        0.000000   2001.000000      0.000000                  0.000000   
25%     1009.000000   2004.000000      8.000000                  6.000000   
50%     3678.000000   2008.000000     22.000000                 20.000000   
75%     6347.000000   2011.000000     44.000000                 49.000000   
max     9016.000000   2014.000000   5076.000000              10626.000000   

       Dowry Deaths  Assault on women with intent to outrage her modesty  \
count  10677.000000                                       10677.000000     
mean      20.181699                                         113.539196     
std       98.276531                                         458.903951     
min        0.000000                                           0.000000     
25%        1.000000                                          10.000000     
50%        5.000000                                          34.000000     
75%       16.000000                                          85.000000     
max     2469.000000                                       10001.000000     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
count                10677.000000                         10677.000000   
mean                    27.419313                           209.224314   
std                    167.806797                           905.664362   
min                      0.000000                             0.000000   
25%                      0.000000                            11.000000   
50%                      2.000000                            50.000000   
75%                     12.000000                           144.000000   
max                   4970.000000                         23278.000000   

       Importation of Girls  
count          10677.000000  
mean               0.175330  
std                2.228637  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max               83.000000

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





1.   **mean**	3871.570198	2007.697949	57.989885	69.888358	20.181699	113.539196	27.419313	209.224314	0.175330
2.   **min**	0.000000	2001.000000	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000
**max**	9016.000000	2014.000000	5076.000000	10626.000000	2469.000000	10001.000000	4970.000000	23278.000000	83.000000



##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [ ]:
df['Insult to modesty of Women'] = df['Insult to modesty of Women'].astype(float)

df.DISTRICT = df.DISTRICT.astype("category")

df.Rape= df.Rape.astype("int")
df.info()
#Вывести типы колонок тут



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10677 entries, 0 to 10676
Data columns (total 11 columns):
 #   Column                                               Non-Null Count  Dtype   
---  ------                                               --------------  -----   
 0   Unnamed: 0                                           10677 non-null  int64   
 1   STATE/UT                                             10677 non-null  object  
 2   DISTRICT                                             10677 non-null  category
 3   Year                                                 10677 non-null  int64   
 4   Rape                                                 10677 non-null  int64   
 5   Kidnapping and Abduction                             10677 non-null  int64   
 6   Dowry Deaths                                         10677 non-null  int64   
 7   Assault on women with intent to outrage her modesty  10677 non-null  int64   
 8   Insult to modesty of Women                           106

In [ ]:
DISTRICT категориальный, Assault on women with intent to outrage her modesty числовой количественный, Dowry Deaths числовой,Rape числовой,
 в моем датасете нет прям отдельных данных, которые можно было бы назвать порядковыми, поэтому Unnamed: 0
 	я возьму как числовой-порядковый

Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [ ]:
#Фильтр колонки 1
df[df["Year"] > 2008]

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
5869         5869  ANDHRA PRADESH           ADILABAD  2009    67   
5870         5870  ANDHRA PRADESH          ANANTAPUR  2009    22   
5871         5871  ANDHRA PRADESH           CHITTOOR  2009    33   
5872         5872  ANDHRA PRADESH           CUDDAPAH  2009    27   
5873         5873  ANDHRA PRADESH          CYBERABAD  2009    82   
...           ...             ...                ...   ...   ...   
10672         832     Lakshadweep        Lakshadweep  2014     1   
10673         833     Lakshadweep  Total District(s)  2014     1   
10674         834      Puducherry           Karaikal  2014     3   
10675         835      Puducherry         Puducherry  2014     7   
10676         836      Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
5869                         56            13   
5870                         60            14   
5871                         46            19   
5872                         47            17   
5873                        113            46   
...                         ...           ...   
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
5869                                                 192     
5870                                                 157     
5871                                                 120     
5872                                                 165     
5873                                                 173     
...                                                  ...     
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
5869                          163                                  428   
5870                           13                                  143   
5871                          268                                  333   
5872                           52                                  124   
5873                          155                                  991   
...                           ...                                  ...   
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
5869                      0  
5870                      0  
5871                      0  
5872                      0  
5873                      0  
...                     ...  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0  

[4808 rows x 11 columns]

In [ ]:
#Фильтр колонки 2
df[df["Insult to modesty of Women"] < 120]

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
0               0  ANDHRA PRADESH           ADILABAD  2001    50   
1               1  ANDHRA PRADESH          ANANTAPUR  2001    23   
2               2  ANDHRA PRADESH           CHITTOOR  2001    27   
3               3  ANDHRA PRADESH           CUDDAPAH  2001    20   
4               4  ANDHRA PRADESH      EAST GODAVARI  2001    23   
...           ...             ...                ...   ...   ...   
10672         832     Lakshadweep        Lakshadweep  2014     1   
10673         833     Lakshadweep  Total District(s)  2014     1   
10674         834      Puducherry           Karaikal  2014     3   
10675         835      Puducherry         Puducherry  2014     7   
10676         836      Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
0                            30            16   
1                            30             7   
2                            34            14   
3                            20            17   
4                            26            12   
...                         ...           ...   
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
0                                                    149     
1                                                    118     
2                                                    112     
3                                                    126     
4                                                    109     
...                                                  ...     
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                              34                                  175   
1                              24                                  154   
2                              83                                  186   
3                              38                                   57   
4                              58                                  247   
...                           ...                                  ...   
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  
...                     ...  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0  

[10273 rows x 11 columns]

In [ ]:
#Фильтр колонки 3
df[df["Dowry Deaths"] > 30]

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
8               8  ANDHRA PRADESH         KARIMNAGAR  2001    56   
14             14  ANDHRA PRADESH           NALGONDA  2001    35   
18             18  ANDHRA PRADESH        RANGA REDDY  2001    72   
26             26  ANDHRA PRADESH           WARANGAL  2001    53   
28             28  ANDHRA PRADESH              TOTAL  2001   871   
...           ...             ...                ...   ...   ...   
10622         782     West Bengal            Hooghly  2014    45   
10631         791     West Bengal        Murshidabad  2014   252   
10632         792     West Bengal              Nadia  2014   106   
10640         800     West Bengal  Total District(s)  2014  1466   
10671         831        Delhi UT  Total District(s)  2014  2096   

       Kidnapping and Abduction  Dowry Deaths  \
8                            49            62   
14                           19            31   
18                           83            37   
26                           25            52   
28                          765           420   
...                         ...           ...   
10622                       213            34   
10631                       457            36   
10632                       336            50   
10640                      4976           501   
10671                      4034           153   

       Assault on women with intent to outrage her modesty  \
8                                                    414     
14                                                   188     
18                                                   113     
26                                                   241     
28                                                  3544     
...                                                  ...     
10622                                                222     
10631                                                523     
10632                                                181     
10640                                               5670     
10671                                               4322     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
8                              81                                  224   
14                             59                                  214   
18                             55                                  421   
26                              0                                  165   
28                           2271                                 5791   
...                           ...                                  ...   
10622                           6                                 1338   
10631                          16                                 2439   
10632                           8                                 1827   
10640                         543                                23278   
10671                        1361                                 3173   

       Importation of Girls  
8                         0  
14                        0  
18                        7  
26                        0  
28                        7  
...                     ...  
10622                     0  
10631                     0  
10632                     1  
10640                     4  
10671                     0  

[1184 rows x 11 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [3]:
#df.groupby(["DISTRICT"]).mean()
df.groupby(['DISTRICT']).mean(numeric_only=True)
#df["DISTRICT"].mean()

Unnamed: 0    Year        Rape  Kidnapping and Abduction  \
DISTRICT                                                                       
24 PARGANAS NORTH  4426.000000  2007.0  153.692308                203.692308   
24 PARGANAS SOUTH  4427.000000  2007.0  185.384615                193.615385   
A and N ISLANDS    5080.000000  2007.0    3.000000                  9.000000   
ADILABAD           3725.384615  2007.0   56.384615                 47.461538   
AGAR                355.000000  2013.0   24.000000                  8.000000   
...                        ...     ...         ...                       ...   
Yamunanagar         213.000000  2014.0   50.000000                 82.000000   
Yavatmal            450.000000  2014.0  108.000000                135.000000   
ZUNHEBOTO          4185.846154  2007.0    1.384615                  0.461538   
ZZ TOTAL            483.228571  2013.0  963.057143               1482.314286   
Zunheboto           495.000000  2014.0    3.000000                  0.000000   

                   Dowry Deaths  \
DISTRICT                          
24 PARGANAS NORTH     55.230769   
24 PARGANAS SOUTH     53.153846   
A and N ISLANDS        1.000000   
ADILABAD              15.384615   
AGAR                   1.000000   
...                         ...   
Yamunanagar            7.000000   
Yavatmal              12.000000   
ZUNHEBOTO              0.000000   
ZZ TOTAL             230.942857   
Zunheboto              0.000000   

                   Assault on women with intent to outrage her modesty  \
DISTRICT                                                                 
24 PARGANAS NORTH                                         171.230769     
24 PARGANAS SOUTH                                         126.923077     
A and N ISLANDS                                            21.000000     
ADILABAD                                                  152.461538     
AGAR                                                       41.000000     
...                                                              ...     
Yamunanagar                                                76.000000     
Yavatmal                                                  466.000000     
ZUNHEBOTO                                                   0.000000     
ZZ TOTAL                                                 2021.114286     
Zunheboto                                                   0.000000     

                   Insult to modesty of Women  \
DISTRICT                                        
24 PARGANAS NORTH                    4.230769   
24 PARGANAS SOUTH                    1.615385   
A and N ISLANDS                      2.000000   
ADILABAD                           106.153846   
AGAR                                 1.000000   
...                                       ...   
Yamunanagar                          0.000000   
Yavatmal                           504.000000   
ZUNHEBOTO                            0.000000   
ZZ TOTAL                           359.685714   
Zunheboto                            0.000000   

                   Cruelty by Husband or his Relatives  Importation of Girls  
DISTRICT                                                                      
24 PARGANAS NORTH                          1474.000000              0.615385  
24 PARGANAS SOUTH                          1588.461538              0.230769  
A and N ISLANDS                              18.000000              0.000000  
ADILABAD                                    353.076923              0.000000  
AGAR                                         30.000000              0.000000  
...                                                ...                   ...  
Yamunanagar                                 170.000000              0.000000  
Yavatmal                                    301.000000              0.000000  
ZUNHEBOTO                                     0.000000              0.000000  
ZZ TOTAL                                   3396.17142

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация